Issue #2313 grids in dynamicfieldsets

chzauleck opened the pull request.

chzauleck committed:
Issue #2313: Added grid functionality to dynamic field sets.

chzauleck committed:
Issue #2313: Added grid functionality to dynamic field sets.

chzauleck committed:
Issue #2313: Added grid functionality to dynamic field sets.

chzauleck committed:
Issue #2313: Added grid functionality to dynamic field sets.

chzauleck committed:
Issue #2313: Added grid functionality to dynamic field sets.

chzauleck committed:
Issue #2313: Added grid functionality to dynamic field sets.

chzauleck committed:
Issue #2313: Layout and general CSS for grids in sets

added css to allow for nicer layout of grids in sets
removed max width of content column to allow for enough space for wide tables

chzauleck committed:
Issue #2313: Layout and general CSS for grids in sets

added css to allow for nicer layout of grids in sets
removed max width of content column to allow for enough space for wide tables

chzauleck committed:
Issue #2313: Layout and general CSS for grids in sets

added css to allow for nicer layout of grids in sets
removed max width of content column to allow for enough space for wide tables

stefanhaerter committed:
Issue #2042: Prevent modal dialog from closing when selecting an autocomplete suggestion.

stefanhaerter committed:
Issue #2042: Added explanatory comment for CloseOnClickOutside special case.

stefanhaerter committed:
Issue #2804: Introduced necessary set adaptions into PostValueSet function of df backend.
Removed object name when setting values inside a set as it is given by the set value itself.

stefanhaerter committed:
Issue #2804: Adapted BaseEntity value structure for usage in sets.

stefanhaerter committed:
Issue #2804: Enabled set-usage for reference fields.

stefanhaerter committed:
Issue #2804: Expanded ValueSetGet unit test Set field coverage.

chzauleck committed:
Issue #2313: multivalue button positioning fixed

chzauleck committed:
Issue #2313: multivalue button positioning fixed

chzauleck committed:
Merge branch 'issue-#2313-grids-in-dynamicfieldsets' of https://github.com/RotherOSS/otobo into issue-#2313-grids-in-dynamicfieldsets

# Conflicts:
#	var/httpd/htdocs/skins/Customer/default/css/Core.InputFields.css

chzauleck committed:
Issue #2313: fixed code policy violations

chzauleck committed:
Issue #2313: fixed code policy violations

chzauleck committed:
Issue #2313: fixed code policy violations

chzauleck committed:
Issue #2313: more code policy fixes

chzauleck committed:
Issue #2313: more code policy fixes

bschmalhofer committed:
Update POD. CustomerUser is not its own field class.

bschmalhofer committed:
Convert test script to Test2::V0

stefanhaerter committed:
Issue #2042: Aligned html structure of reference search field with normal field to improve functionality.

stefanhaerter committed:
Issue #2042: Increased z-index of autocomplete dropdowns to make them visible when used in overlay dialogs.

stefanhaerter committed:
Issue #2042: Enable agent reference search module to work with changes to reference search field structure.

stefanhaerter committed:
Issue #2042: Enhanced comment for css change.

stefanhaerter committed:
Issue #2804: Included df set in EditFieldValueGet unit test.

stefanhaerter committed:
Issue #2804: POD correction.

stefanhaerter committed:
Issue #2804: Added sorting for df reference select dropdowns.

stefanhaerter committed:
Issue #2804: Added case for df set to EditFieldRender unit test.

stefanhaerter committed:
Issue #2804: Re-aligned template toolkit variables to prevent whitespaces and newlines which produce confusing unit test errors.

stefanhaerter committed:
Issue #2804: Rerolled previous change regarding whitespaces and adapt unit test instead.

stefanhaerter committed:
Issue #2804: Added forgotten indentation.

svenoe commented on Kernel/Modules/AdminDynamicFieldSet.pm:
```diff
@@ -137,6 +137,9 @@ sub _AddAction {
 
+    my $DynamicFieldObject        = $Kernel::OM->Get('Kernel::System::DynamicField');
+    my $DynamicFieldBackendObject = $Kernel::OM->Get('Kernel::System::DynamicField::Backend');
+
```
Why was this shifted further up?

svenoe commented on Kernel/Modules/AdminDynamicFieldSet.pm:
```diff
@@ -148,48 +151,97 @@ sub _AddAction {
+                if ( !$DynamicField ) {
                     $Errors{IncludeServerError}        = 'ServerError';
-                    $Errors{IncludeServerErrorMessage} = Translatable( 'No valid dynamic field "' . $Line->{DF} . '".' );
+                    $Errors{IncludeServerErrorMessage} = Translatable( 'No dynamic field "' . $DFElement . '".' );
```
Field could exist but be set to invalid, I would leave the "valid".

svenoe commented on Kernel/Modules/AdminDynamicFieldSet.pm:
```diff
@@ -148,48 +151,97 @@ sub _AddAction {
+
+                        for my $DFEntry ( $Row->@* ) {
+
+                            if ( !$CheckDFElement->($DFEntry->{DF}) ) {
+                                last LINE;
+                            }
+                        }
+                        
+                    }
+
+                    push @Include, { Grid => $Line->{Grid} };
+
+                } elsif ( $Line->{DF} ) {
```
Not critical, but also not 100% consistent. For now, I would force that each $Line contains either a grid or a DF, thus change `elsif ( $Line->{DF} )` to a simple `else`. Also it would be nice to somewhere (probably in `&$CheckDFElement`) catch the possibility that e.g. `$DFEntry->{DF}` is `undef`.

chzauleck replied:
wouldnt it be nicer to leave the elsif and add error handling in an else clause?

svenoe commented on Kernel/Modules/AdminDynamicFieldSet.pm:
```diff
@@ -430,38 +482,90 @@ sub _ChangeAction {
         );
 
         if ( IsArrayRefWithData($IncludeFrontend) ) {
```
As the whole check is done at two places and it grew a bit, one could think about moving the complete content of this block into a sub. The only "downside" is, that `%Errors` should not be altered in there directly, in this case and things would have to be slightly rewritten.
```
if ( IsArrayRefWithData($IncludeFrontend) ) {
    my %YAMLErrors = $Self->_CheckInclude( $IncludeFrontend );
    ...
}
```

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
+# $Include: a list of hash references containing information about the items to include
+# $DynamicFieldObject: an object used to retrieve dynamic field information
+# and returns either the DynamicFields or undef in case of an error.
+my $GetIncludedDynamicFields = sub {
```
While this is possible, I have not seen anonymous subs used outside of subs anywhere else in OTOBO, and I'm not sure why you would want that here. If there is no real reason, please rewrite it to:
```
sub _GetIncludedDynamicFields {
...
}
```
Also take a look at other places on how to write pod, e.g. https://github.com/RotherOSS/otobo/blob/40a68a17babe7254279c3f27302f037ad8129269/Kernel/System/Ticket.pm#L8003-L8015 or https://github.com/RotherOSS/otobo/blob/40a68a17babe7254279c3f27302f037ad8129269/Kernel/System/Console/BaseCommand.pm#L881-L891

(Actually we could decide to consistently use one of the two versions, but this is not and has not been in a consistent state.)

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
+        return $DynamicField;
+    };
+
+    my $Error = undef;
```
As Bernhard noted, this is equivalent to just `my $Error;`.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
+# and returns either the DynamicFields or undef in case of an error.
+my $GetIncludedDynamicFields = sub {
+
+    my ($Include, $DynamicFieldObject, $DynamicFieldName) = @_;
```
In most cases this should be passed as, then,
```
my $DynamicFields = _GetIncludedDynamicFields(
    Include => $Include,
    DynamicFieldName => $DynamicFieldName,
    DynamicFieldObject => $DynamicFieldObject,
);
```
and **could** be checked similar to e.g. https://github.com/RotherOSS/otobo/blob/40a68a17babe7254279c3f27302f037ad8129269/Kernel/System/Ticket.pm#L420-L432
(in this case probably not necessary, just as comment to note; in general for internal functions this is less important and rarely done)

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
+
+            for my $Row (@{ $IncludeItem->{Grid}{Rows} }) {
+
+                for my $DFEntry (@$Row) {
```
This is also rarely to never used. Please rather use either "the old way" `@{ $Row }`, or alternatively `$Row->@*` (which does the same).

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
+    # Returns the dynamic field definition
+    my $GetDynamicField = sub {
+
+        my ($DFEntry, $DynamicFieldObject) = @_;
```
As you are defining the sub within the scope of `GetIncludedDynamicFields` and `$DynamicFieldObject` is already initialized in line 92, you do not need to pass it to `$GetDynamicField`.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
+    if ($Error) {
+        $Kernel::OM->Get('Kernel::System::Log')->Log(
+            Priority => 'error',
+            Message  => "$DynamicFieldName configured erroneously. 'DF: DynamicFieldName' missing or wrong.",
```
You could set `$Error` to either "No DynamicField '$IncludeItem->{DF}'.", or "Need either Grid or DF in each Row.", or similar, based on the context, and then set the Message to `"$DynamicFieldName configured erroneously: $Error"` - in most cases probably not too important, as the "DynamicFieldGet()" would throw an error with the name of the field, but for missing Grid or DF probably more clear.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
-                Priority => 'error',
-                Message  => "$Param{DynamicFieldConfig}{Name} configured erroneously. 'DF: DynamicFieldName' missing or wrong.",
-            );
+    #DEBUG
+    #use Data::Dumper;
+    #print STDERR "DynamicFields".Dumper($DynamicFields)."\n";
```
Please remove.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -81,28 +81,98 @@ sub new {
 
-            return;
-        }
+    for my $i (0 .. $#$DynamicFields) {
```
Also here please use `$#{ $DynamicFields }`.

stefanhaerter committed:
Issue #2785: Fixed typos in comments.

stefanhaerter committed:
Issue #2785: Aligned article dynamic field value behavior with other screens to enable usage of default values.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -130,37 +200,33 @@ sub ValueSet {
+    my $DynamicFieldName = $Param{DynamicFieldConfig}{Name};
+    my $DynamicFields = $GetIncludedDynamicFields->($Include, $DynamicFieldObject, $DynamicFieldName);
+
+    if (!$DynamicFields) {
+        
+        $Success = 0;
+    } else {
```
Please `return;` instead of `return 0;` or `return undef;` or so, in case of errors. (With the exception of something like `return { Success => 0, Error => 'Did not work' };` where you explicitely want to return more information even in case of an error.)

Here the usual way would be:
```
return if !$DynamicFields;

for my $i ( 0 .. $#{ $DynamicFields } ) {
    ...
```

(One could even argue whether a ValueSet() for a set which has no dynamic fields in it should just `return 1;`, actually, but it's a more theoretical question, I guess...)

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -130,37 +200,33 @@ sub ValueSet {
+        
+        for my $i (0 .. $#$DynamicFields) {
+
+            my $DynamicField = $DynamicFields->[$i];
```
For my personal taste a bit much, but not wrong ofc.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -130,37 +200,33 @@ sub ValueSet {
-        {
-            $Success = 0;
+            {
+                $Success = 0;
```
The question is whether you even want to try to continue updating the other fields. One maybe could also just return here. Leave a log message though, please. See e.g. https://github.com/RotherOSS/otobo/blob/40a68a17babe7254279c3f27302f037ad8129269/Kernel/System/Ticket.pm#L426-L430
(With an added blank line in front of the return, please.)

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -453,31 +511,31 @@ sub EditFieldValueValidate {
     my $Result;
     for my $i ( 0 .. $#{ $Param{DynamicFieldConfig}{Config}{Include} } ) {
```
This looks wrong to me, should have to be deleted, too, I think.

svenoe commented on Kernel/System/DynamicField/Driver/Set.pm:
```diff
@@ -453,31 +511,31 @@ sub EditFieldValueValidate {
+        } else {
+            for my $i (0 .. $#{$DynamicFields}) {
+
+            my $DynamicField = $DynamicFields->[$i];
+                # prevent overwriting names in cached data
+                $DynamicField = { $DynamicField->%* };
```
Please combine... ;)

svenoe commented on var/httpd/htdocs/skins/Agent/default/css/Core.Form.css:
```diff
@@ -707,31 +707,41 @@ fieldset.TableLike > .Row.MultiColumn > .FieldCell > .Field {
     width: 100%;
     box-sizing: border-box;
+    text-align: left;
+    text-overflow: clip;
```
Isn't ellipsis nicer?

svenoe commented on var/httpd/htdocs/skins/Agent/default/css/Core.InputFields.css:
```diff
@@ -345,8 +353,8 @@ div.InputField_ToolbarContainer ul li {
     content: "|";
-    display: inline-block;
-    float: right;
+    /*display: inline-block;*/ /*inline block is ignored due to float*/
```
Then please remove it completely.

svenoe commented on var/httpd/htdocs/skins/Customer/default/css/Core.InputFields.css:
```diff
@@ -1067,7 +1110,7 @@ div.InputField_ToolbarContainer ul li a:hover {
 
 .jstree-InputField-Tree .jstree-icon {
     font-family: 'otobo' !important;
-    speak: none;
+    /*speak: none; /*unknown property?!*/
```
Please research and either leave it in, or remove it completely.

svenoe reviewed (changes requested):
In general it looks good, please use the code policy, though (I left some stuff uncommented, where I think, that it will tidy your code a bit anyways). For the Set driver, I did not add comments in places where stuff repeated. In the beginning I wrote quite a lot, also small stuff, just to say as much as possible, further down I looked through the code a bit faster. Especially the css changes are more or less unchecked, as I didn't look at the changes in the frontend. We can make adaptions there, later, too.

Stefan recently added some unit tests specifically for sets. Please ask him which and run them with your changes.

bschmalhofer committed:
Enhance POD of the method UserAdd().

stefanhaerter committed:
Issue #2812: Fixed error handling of yaml string in AdminDynamicFieldSet.

bschmalhofer committed:
Issue #2813: Tidying, improve POD and comments

bschmalhofer committed:
Issue #2813: tentatively add a JSON formatter

Add the formatter.
Add the method FormatterCanHandleReferences() because data preparation
depends on the capabilities of the formatter backend.
Do not activate per default as the export writes one
object after the other, which does not form a valid JSON file.

bschmalhofer committed:
Issue #2817: passing no arguments also prints the usage

bschmalhofer committed:
Issue #2817: sync with the code in rel-11_0

There are no functional changes. The aim is just to make
the code more similar, so that maintainance is easier.

bschmalhofer committed:
Issue #2817: removed redundant 'Required => 0' declarations

Whethe a module is installed depends on the feature, when
the feature of the module has been declared.

bschmalhofer committed:
Issue #2817: remove an unused variable

bschmalhofer committed:
Issue #2817: require at least DBD::mysl 4.00

Just to have a nominal minimum version. DBD::mysql 4.00 was released in 2006.

bschmalhofer committed:
Issue #2817: Blacklist DBD::mysql 4.042, 5.002, 5.003

5.001 was already blacklisted.
This fixes the current problem that the Docker image can't be built.

bschmalhofer committed:
Merge pull request #2819 from RotherOSS/issue-#2817-dbd_mysql_ge_5_001

Issue #2817 dbd mysql ge 5 001

bschmalhofer committed:
Merge branch 'rel-10_0' into rel-10_1

bschmalhofer committed:
Selectively apply some changes from rel-10_1

Just to make merging more simple

bschmalhofer committed:
Merge branch 'rel-10_1' into rel-11_0

With help from Stefan Härter.

bschmalhofer committed:
Merge pull request #2821 from RotherOSS/issue-#2813-importexport_json

Issue #2813 importexport json

bschmalhofer committed:
Issue #2820: reorganize the ImportExport test scripts

That is, put them into scripts/test/ImportExport

bschmalhofer committed:
Merge pull request #2822 from RotherOSS/issue-#2820-scripts_test_importexport

Issue #2820: reorganize the ImportExport test scripts

bschmalhofer committed:
Issue #2788: Saner description of setting DefaultLanguage.

Referring to the next section is not sensible in the SysConfig GUI

bschmalhofer committed:
Issue #2788: replace 'front-end' with 'frontend'

in the description of SysConfig settings. 'front-end' is not really incorrect,
but 'frontend' is more commonly used.

bschmalhofer committed:
Merge pull request #2823 from RotherOSS/issue-#2788-default_language

Issue #2788 default language

bschmalhofer committed:
Issue #2824: update the copyright year

by running: ../CodePolicy_10_0/bin/otobo.CodePolicy.pl --plugins +TidyAll::Plugin::OTOBO::Legal::ReplaceCopyright -a .

bschmalhofer committed:
Issue #2824: accept suggestins from CodePolicy

even when the don't really improve the code

bschmalhofer committed:
Issue #2824: Set copyright year to 2019-2024

previously there was only the specific year 2021 set and
.screenrc was not included for the ReplaceCopyright code policy.

bschmalhofer committed:
Issue #2824: adapt copyright year

Apparently Kernel/Output/HTML/Templates/Standard/CustomerFooter.tt
and Kernel/System/MigrateFromOTRS/Base.pm are not handled by the
OTOBO/Legal/ReplaceCopyright policy.

bschmalhofer committed:
Merge pull request #2825 from RotherOSS/issue-#2824-copyright_2024

Issue #2824 copyright 2024

bschmalhofer committed:
Merge branch 'rel-10_0' into rel-10_1

Update of the copyright year

bschmalhofer committed:
Update copyright year

for files that are not in rel-10_0

bschmalhofer committed:
Accept suggestions from CodePolicy

bschmalhofer committed:
Merge branch 'rel-10_1' into rel-11_0

bschmalhofer committed:
Update copyright year to 2019-2024

for the files that were not in rel-10_1

bschmalhofer committed:
Accept suggestion from CodePolicy.

Remove trailing whitespace.

bschmalhofer committed:
Accept suggestions from CodePolicy

This is the first round using Perl::Tidy 20230912.
The changes do make sense, in my opinion.
The major part of the changes are the alignment of pararmeters when
'use', 'push', or 'reverse' statements are on consecutive lines.

bschmalhofer committed:
Update copyright year in sample file.

This surfaced because there was an unittest failure
in scripts/test/Loader.t

Commit:
Issue #2814: Fix value set for base entities in DFSet context.

Commit:
Issue #2814: Fix EditFieldRender bug with existing values. Add small usage note.

stefanhaerter committed:
Issue #2827: Copied frontend admin regex code from text to script.

stefanhaerter committed:
Issue #2827: Added regex evaluation to base script df backend.

stefanhaerter committed:
Issue #2827: Added todo comments.

bschmalhofer committed:
Issue #2898: use Module::Metadata->find_module_by_name()

for determining whether a module is installed when '--inst' was passed.
Mostly for using the standard toolchain methods.

bschmalhofer committed:
Issue #2898: use CPAN::Meta::Prereqs->accept_module()

for checking the validity of a module version.
This obsoletes the methed CleanVersion().
accept_module() also support blacklisted modules

bschmalhofer committed:
Issue #2898: simplify the generation of the cpanfile

because the version range is now already available in the desired format.

bschmalhofer committed:
Issue #2898: VersionNotSupported is now part of VersionRequired.

For DBD::mysql accept no versions >= 5.001.
This was the initial trigger for this issue.

bschmalhofer committed:
Issue #2898: Template::Stash::XS declares no version

So provide a fallback for the output.

bschmalhofer committed:
Issue #2898: specify the '>=' operator in VersionRequired

This is not really necessary, as '>=' is the default. But adding the operator
is nice, because is was previously already part of the generated cpanfiles.

bschmalhofer committed:
Issue #2898: regenerate the cpanfiles

The delimiters of the version ranges have change from double to single quotes.
Apparently both are valid.

bschmalhofer committed:
Issue #2898: satisfy Code::Policy 10.0

with the Perl::Tidy version from 2019

bschmalhofer committed:
Merge pull request #2829 from RotherOSS/issue-#2898-module_metadata_take2

Issue #2898 module metadata take2

bschmalhofer committed:
Merge branch 'rel-10_0' into rel-10_1

bschmalhofer committed:
Update code comment. VersionsNotSupported is gone.

This aattribute is now included in VersionRequired.

bschmalhofer committed:
Merge branch 'rel-10_0' into rel-10_1

bschmalhofer committed:
Issue #2898: a saner way of setting up VersionRequired

No need to use CPAN::Meta::Prereqs

bschmalhofer committed:
Merge pull request #2830 from RotherOSS/issue-#2898-add_string_requirement

Issue #2898: a saner way of setting up VersionRequired

bschmalhofer committed:
Merge branch 'rel-10_0' into rel-10_1

bschmalhofer committed:
Manually update cpanfile.docker.kerberos.

This should no longer be necessary in rel-11.0

bschmalhofer committed:
Merge branch 'rel-10_1' into rel-11_0

bschmalhofer committed:
Regenerate cpan-lib/cpanfile.

Only the delimiters of the version range changed from
double quotes to single quotes.

stefanhaerter committed:
Passed dynamic field config as param for ObjectDescriptionGet.
This is needed because there is no possibility to get DisplayType value inside the function.

stefanhaerter committed:
Offer different display types for ticket reference field.

bschmalhofer committed:
Issue #2898: eliminate the special case for ExactVersion

Use the regular VersionRequired instead.

bschmalhofer committed:
Issue #2813: formatting option for JSON pretty print

Pretty per default

bschmalhofer committed:
Issue #2813: file extension and MIME type

for concatenated JSON export.
This is only tentative.

bschmalhofer committed:
Merge pull request #2832 from RotherOSS/issue-#2813-json_pretty

Issue #2813 json pretty

stefanhaerter committed:
Issue #2833: Added function to get ticket attribute hash.

stefanhaerter committed:
Issue #2833: Added function to get agent attribute hash.

bschmalhofer committed:
Issue #2498: open the otobo.de links in new tab.

These are the links at the bottom of the customer interface.
The commented out FooterLinks already open in a new tab.

bschmalhofer committed:
Merge pull request #2834 from RotherOSS/issue-#2498-customer_footer

Issue #2498: open the otobo.de links in new tab.

bschmalhofer committed:
Merge branch 'rel-10_0' into rel-10_1

bschmalhofer committed:
Merge branch 'rel-10_1' into rel-11_0

stefanhaerter committed:
Issue #2833: Slightly reworked ticket ObjectMappingGet sub.

stefanhaerter committed:
Issue #2833: Added extended attributes to ticket ObjectMappingGet sub.

stefanhaerter committed:
Issue #2833: Renamed functions and corrected POD.

stefanhaerter committed:
Issue #2833: Added default value for param Extended.
Re-ordered handling of sysconfig-depending attributes.
Removed falsly included attributes.

stefanhaerter committed:
Issue #2833: Improved variable names.

bschmalhofer committed:
Issue #2846: Eliminate unused variable $EncodeObject

bschmalhofer committed:
Issue #2846: return the content type application/json

just like the AgentCustomerSearch.pm frontend does.

bschmalhofer committed:
Issue #2846: tidying

Make Kernel/Modules/AgentCustomerSearch.pm and Kernel/Modules/AgentUserSearch.pm
more similar to each other.

chzauleck committed:
Issue #2313: minor css code cleaning

removed unused css statements
changed clip to ellipsis for labels in dfsets

chzauleck committed:
Issue #2313: minor css code cleaning

removed unused css statements
changed clip to ellipsis for labels in dfsets

chzauleck committed:
Issue #2313: code cleaning

improved error handling and coding style consistency

chzauleck committed:
Issue #2313: code cleaning

improved error handling and coding style consistency

chzauleck committed:
Issue #2313: code cleaning

improved error handling and coding style consistency

bschmalhofer committed:
Merge pull request #2837 from RotherOSS/issue-#2846-application_json

Issue #2846 application json

bschmalhofer committed:
Enhance the POD of the method TestUserCreate().

In list context the user ID is returned as well

stefanhaerter committed:
#2841:  Dropdowns for dynamic field reference filters in AdminDynamicFieldReference

* Changed input fields for object attributes in dynamic reference filter rows to dropdowns.

* Fixed typo.

chzauleck committed:
Merge branch 'issue-#2313-grids-in-dynamicfieldsets' of https://github.com/RotherOSS/otobo into issue-#2313-grids-in-dynamicfieldsets

chzauleck committed:
Merge branch 'issue-#2313-grids-in-dynamicfieldsets' of https://github.com/RotherOSS/otobo into issue-#2313-grids-in-dynamicfieldsets
